Refactor MiqRequestTask::StateMachine and MiqRequestTask::StateMachine specs

djberg96 opened the pull request:
Currently the `MiqRequestTask` state machine specs fail with the following error if strict partial double validation is turned on:

```
Failure/Error: allow(task).to receive(:some_state).and_raise(exception)
       #<MiqRequestTask id: 72000000000110, description: nil, state: "pending", request_type: nil, userid: nil, options: {}, created_on: "2020-02-03 19:22:44", updated_on: "2020-02-03 19:22:44", message: nil, status: "Ok", type: nil, miq_request_id: nil, source_id: nil, source_type: nil, destination_id: nil, destination_type: nil, miq_request_task_id: nil, phase: nil, phase_context: {}, tenant_id: nil, cancelation_status: nil, conversion_host_id: nil> does not implement: some_state
```

This initial error can be solved by switching `receive(:some_state)` with `receive(:send).with(:some_state)` which is what it's doing internally.

The overall logic is somewhat confusing. Near as I can tell it follows this workflow:

* `MiqRequestTask#signal(:some_state)` bypasses the initial `if` check since it is not `prematurely_finished?`
* It calls `save`, which triggers `set_tenant`
* A `send(:some_state)` call is made
* Since there is no `some_state` method, this triggers a `NoMethodError`, which is caught and then `signal(:provision_error)` is called
* Now back at the top of the `signal` method, it bypasses the `if` check again since the phase is now `:provision_error`
* The `save` method is called again, which triggers `set_tenant` again
* A `send(:provision_error)` call is made, which calls the `provision_error` method, which in turn calls `update_and_notify_parent`
* The `update_and_notify_parent` method bombs on the `parent.reload` method because the parent is nil
* The above error is inadvertently rescued, and the error and backtrace is logged. The end.
* But, let's assume the parent isn't nil and the `update_and_notify_parent` works as expected, it would then call `signal(:finish)` at the end of the `provision_error` method.
* Now back at the top of the `signal` method AGAIN, it bypasses the `if` check again since the phase is now `:finish`
* Another call to `save` and thus `set_tenant`
* It tries to call `send(:finish)`, which of course fails since there is no `finish` method, which is again rescued, logged, and finally that's the end of it.

You can see this for yourself:

* Insert some debugging statements like so: https://gist.github.com/djberg96/47c743ff42caec71f90cad57b053d6dd
* Also add a debug print inside the `set_tenant` method within the `TenancyMixin` module.
* Fire up rails console.
* Do `t = MiqRequestTask.new` (or `t = FactoryBot.create(:miq_request_task)` if you prefer test mode)
* Call `t.signal(:some_state)`, and watch workflow.

My other problem with this code is that it isn't really a state machine. Not what I would consider one, anyway.

Fryguy commented:
Can you do `receive(:signal).with(:some_state).and_call_original`?  or does it need to stub that away?

djberg96 commented:
@Fryguy You mean on line 14? If I do that I get:

```
received :send with unexpected arguments
  expected: (:some_state), got: (:set_tenant)
```


If I also stub the call to `.save`, then we're back to the current error.

kbrock commented on spec/models/miq_request_task/state_machine_spec.rb:
```diff
@@ -1,15 +1,19 @@
         allow(task).to receive_messages(:miq_request => double("MiqRequest").as_null_object)
         exception = String.xxx rescue $!
-        allow(task).to receive(:some_state).and_raise(exception)
+        allow(task).to receive(:send).with(:some_state).and_raise(exception)
```
Wish this were calling `call_original` or something

I wonder if all these `expect` calls are stubbing out `miq_request_task` and so that is why the `task.status` is not properly getting set

kbrock replied:
it seems calling `some_state` would raise an error anyway.
Wonder why we need to `allow` it if it will already raise an exception

djberg96 replied:
@kbrock I think that was done to test the logging.

djberg96 commented:
Update: This ended up being more refactoring than I originally intended. It turns out the original problem that @Fryguy reported way back in 2012 (https://github.com/ruby/psych/issues/85) was fixed a couple years later (https://github.com/ruby/psych/commit/29d4544ff07dbb51421c7fab6e8bd54baecc78c0), which means we can dispense with the special handling altogether.

So, I've removed the `provision_error` method completely, and updated the specs accordingly. I also let the code raise an error naturally instead of explicitly stubbing one, and modified the log formatting slightly so that it includes the error class name.

djberg96 commented:
@miq-bot remove_label wip

djberg96 commented:
@miq-bot add_label test, refactoring

djberg96 commented:
@miq-bot add_reviewer @Fryguy

Fryguy commented on app/models/miq_request_task/state_machine.rb:
```diff
@@ -21,29 +21,15 @@ def signal(phase)
-        _log.log_backtrace(err)
+      case phase
+      when :finish
+        $log.error("[#{err.class}: #{err.message}] encountered during [#{phase}]")
```
Please use _log over $log as it was before.  As such, there is no need for adding the class in

djberg96 replied:
Alright, updated.

Fryguy commented on app/models/miq_request_task/state_machine.rb:
```diff
@@ -21,29 +21,15 @@ def signal(phase)
-        phase_context[:exception_message]   = err.message
-        phase_context[:exception_backtrace] = err.backtrace
-        phase_context[:error_phase]         = phase
-        signal :provision_error
```
I'm confused... Did you just completely eliminate the storing of exceptions in the db altogether?

Additionally, there may have been callers (perhaps via api or automate) that were expecting these calls in the phase context.

djberg96 replied:
Based on the comments, it looked to me like the reason for setting the `phase_context` explicitly like that was no longer necessary and that the `save` method already did it. Specifically, I was looking at this comment:

https://github.com/ManageIQ/manageiq/blob/master/app/models/miq_request_task/state_machine.rb#L75

Are you saying I should explicitly restore these four lines?

```
phase_context[:exception_class]     = err.class.name
phase_context[:exception_message]   = err.message
phase_context[:exception_backtrace] = err.backtrace
phase_context[:error_phase]         = phase
```

Fryguy replied:
>  it looked to me like the reason for setting the phase_context explicitly like that was no longer necessary and that the save method already did it

phase_context is the raw state data passed from state to state...it gets filled in and then save is called to persist it.

Even so, I see that perviously when we rescue the exception, we add it to the phase_context, but then we delete it almost immediately and log it.  Even so, in this else condition we signal provision error and the error caught here is completely lost (at least from what I can tell).

kbrock commented:
> - It calls `save`, which triggers `set_tenant`

Good catch.

In a number of places I added `:if => :tenant_changed?` or `miq_group_changed?`. Since `miq_group` is still our external interface into `tenant`, we need to make sure we give others the ability to modify that.

The `_changed?` allows us to reduce the tenant update calls which are not necessary

> - The above error is inadvertently rescued, and the error and backtrace is logged. The end.

Sure feels like we're not actually testing what we need to test. Rspec complains about `not_to raise_exception(SpecificError)` because another exception could hide the issue. Our code seems to be doing something similar in nature.

Fryguy commented on app/models/miq_request_task/state_machine.rb:
```diff
@@ -21,29 +21,15 @@ def signal(phase)
-        phase_context[:exception_backtrace] = err.backtrace
-        phase_context[:error_phase]         = phase
-        signal :provision_error
+        signal(:provision_error)
```
Where is the provision_error method for this to call, since it was removed below?

djberg96 replied:
Didn't see the point of it. Like you said, it unset everything in the `phase_context` that was set previously (which is why I removed that bit). I removed the logging because ultimately it would result in two log messages (one at `:provision_error`, one on `:finish`) which seemed redundant.

That just leaves calling `update_and_notify_parent` plus `signal(:finish)`. Rather than make a separate method, I just inlined it.

kbrock commented on app/models/miq_request_task/state_machine.rb:
```diff
@@ -21,29 +21,15 @@ def signal(phase)
     rescue => err
-      if [:finish, :provision_error].include?(phase.to_sym)
-        _log.error("[#{err.message}] encountered during [#{phase}]")
+      case phase
```
```suggestion
      case phase.to_sym
```
do we want to keep the `to_sym` here?

djberg96 replied:
@kbrock Done!

djberg96 commented:
I'll let someone else do this.
